### ISSUE TITLE
Elementor and Elementor custom skin integration with POD: POD Not Found error on archive page

### ISSUE THREAD

**Issue by gdeldeo** [pods-framework/pods#5165]
## Issue Overview
When setting up an archive page, Elementor gives you a "Posts" and an "Archive" widget. They both display the posts using 2 standard templates called skins. One can modify the query and therefore make use of custom fields, but none of these two standard templates can show custom fields value. 
In order to create a new Skin one either has to create a class and include it in the function.php or can use a plugin called Elementor Custom Skin.
This plug in basically integrates with elementor and gives the ability to define the skin as desired, hence including all the custom fields that one needs.
When you define the single block that will be repeated as long as there are posts, all the value of the pods are retrieved correctly from the Elementor integration as much as from the single item Pods plugin (which you can embed seamingless in any elementor built). But when the block is actually repeated to build the archive page, elementor is able to retreive the fields belonging to the post type of interest (it cannot go down the relationship and gather the fields of the related post type and so on)
The Pods widget on the contrary fails and returns a Pod not found error. The error is due to the if ! $pod->valid() in your general.php file on row 840

## Expected Behavior
The pods single item widget should represent the items related to each of the posts in the archive page

## Current Behavior
POD not found error

## Steps to Reproduce (for bugs)
1. Create a loop template in elementor 
2. Set the preview to the post of interest 
3. Insert heading widget and assign any dynamic value coming from pods
4. insert the pods single item widget
5. create a new page
6. within the new page insert the posts widget
7. select the custom skin created as loop template

## WordPress Environment
Debug Information
WordPress Version: 4.9.8

PHP Version: 7.2.8

MySQL Version: 5.7.23

Server Software: Apache

Your User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36

Session Save Path: /Applications/MAMP/tmp/php

Session Save Path Exists: Yes

Session Save Path Writeable: Yes

Session Max Lifetime: 1440

Opcode Cache:

Apc: No
Memcached: No
OPcache: No
Redis: No
Object Cache:

APC: No
APCu: No
Memcache: No
Memcached: No
Redis: No
WPDB Prefix: wp_

WP Multisite Mode: No

WP Memory Limit: 40M

Pods Network-Wide Activated: No

Pods Install Location: /Users/peppe/Sites/pods/wp-content/plugins/pods/

Pods Tableless Mode Activated: No

Pods Light Mode Activated: No

Currently Active Theme: Twenty Seventeen

Currently Active Plugins:

Ele Custom Skin: 1.0.5
Elementor: 2.2.4
Elementor Pro: 2.1.9
Pods - Custom Content Types and Fields: 2.7.9

## Pods Package Export (helpful!)
{"meta":{"version":"2.7.9","build":1538221959},"pods":{"10":{"id":10,"name":"avvocati","label":"Avvocati","description":"","type":"post_type","storage":"meta","object":"","alias":"","fields":{"titolo":{"id":11,"name":"titolo","label":"Titolo","description":"","help":"","class":"","type":"text","weight":0,"pick_object":"","pick_val":"","sister_id":"","required":"0","text_allow_shortcode":"0","text_allow_html":"0","text_allowed_html_tags":"strong em a ul ol li b i","text_max_length":"255","pick_post_status":["publish"],"admin_only":"0","restrict_role":"0","restrict_capability":"0","hidden":"0","read_only":"0","roles_allowed":["administrator"],"rest_read":"0","rest_write":"0","rest_pick_response":"array","rest_pick_depth":"2","unique":"0","text_repeatable":"0","text_placeholder":""},"ruolo":{"id":12,"name":"ruolo","label":"Ruolo","description":"","help":"","class":"","type":"pick","weight":1,"pick_object":"custom-simple","pick_val":"","sister_id":"","required":"0","pick_format_type":"single","pick_format_single":"dropdown","pick_format_multi":"checkbox","pick_allow_add_new":"1","pick_taggable":"0","pick_show_icon":"1","pick_show_edit_link":"1","pick_show_view_link":"1","pick_limit":"0","pick_user_role":[],"pick_post_status":["publish"],"admin_only":"0","restrict_role":"0","restrict_capability":"0","hidden":"0","read_only":"0","roles_allowed":["administrator"],"rest_read":"0","rest_write":"0","rest_pick_response":"array","rest_pick_depth":"2","unique":"0","pick_custom":"Partner\nSenior Partner","pick_select_text":"","pick_table_id":"","pick_table_index":"","pick_display":"","pick_where":"","pick_orderby":"","pick_groupby":""},"specialit":{"id":13,"name":"specialit","label":"Specialità","description":"","help":"","class":"","type":"pick","weight":2,"pick_object":"post_type","pick_val":"specialit","sister_id":"16","required":"0","pick_format_type":"multi","pick_format_single":"dropdown","pick_format_multi":"checkbox","pick_allow_add_new":"1","pick_taggable":"0","pick_show_icon":"1","pick_show_edit_link":"1","pick_show_view_link":"1","pick_limit":"3","pick_user_role":[],"pick_post_status":["publish"],"admin_only":"0","restrict_role":"0","restrict_capability":"0","hidden":"0","read_only":"0","roles_allowed":["administrator"],"rest_read":"0","rest_write":"0","rest_pick_response":"array","rest_pick_depth":"2","unique":"0","pick_select_text":"","pick_table_id":"","pick_table_index":"","pick_display":"","pick_where":"","pick_orderby":"","pick_groupby":""},"studio":{"id":14,"name":"studio","label":"Studio","description":"","help":"","class":"","type":"pick","weight":3,"pick_object":"post_type","pick_val":"studi","sister_id":"9","required":"0","pick_format_type":"single","pick_format_single":"dropdown","pick_format_multi":"checkbox","pick_allow_add_new":"1","pick_taggable":"0","pick_show_icon":"1","pick_show_edit_link":"1","pick_show_view_link":"1","pick_limit":"0","pick_user_role":[],"pick_post_status":["publish"],"admin_only":"0","restrict_role":"0","restrict_capability":"0","hidden":"0","read_only":"0","roles_allowed":["administrator"],"rest_read":"0","rest_write":"0","rest_pick_response":"array","rest_pick_depth":"2","unique":"0","pick_select_text":"","pick_table_id":"","pick_table_index":"","pick_display":"","pick_where":"","pick_orderby":"","pick_groupby":""}},"show_in_menu":"1","label_singular":"Avvocato","public":"1","show_ui":"1","supports_title":"1","supports_editor":"1","publicly_queryable":"1","exclude_from_search":"0","capability_type":"post","capability_type_custom":"avvocati","capability_type_extra":"1","has_archive":"0","hierarchical":"0","rewrite":"1","rewrite_with_front":"1","rewrite_feeds":"0","rewrite_pages":"1","query_var":"1","can_export":"1","default_status":"draft","supports_author":"0","supports_thumbnail":"0","supports_excerpt":"0","supports_trackbacks":"0","supports_custom_fields":"0","supports_comments":"0","supports_revisions":"0","supports_page_attributes":"0","supports_post_formats":"0","built_in_taxonomies_category":"0","built_in_taxonomies_elementor_font_type":"0","built_in_taxonomies_elementor_library_type":"0","built_in_taxonomies_link_category":"0","built_in_taxonomies_post_tag":"0","show_in_nav_menus":"1","show_in_admin_bar":"1","pfat_enable":"0","pfat_run_outside_loop":"0","pfat_append_single":"append","pfat_filter_single":"the_content","pfat_append_archive":"append","pfat_filter_archive":"the_content","rest_enable":"0","read_all":"0","write_all":"0"},"15":{"id":15,"name":"specialit","label":"Specialità","description":"","type":"post_type","storage":"meta","object":"","alias":"","fields":{"avvocati":{"id":16,"name":"avvocati","label":"Avvocati","description":"","help":"","class":"","type":"pick","weight":0,"pick_object":"post_type","pick_val":"avvocati","sister_id":"13","required":"0","pick_format_type":"single","pick_format_single":"dropdown","pick_format_multi":"checkbox","pick_allow_add_new":"1","pick_taggable":"0","pick_show_icon":"1","pick_show_edit_link":"1","pick_show_view_link":"1","pick_limit":"0","pick_user_role":[],"pick_post_status":["publish"],"admin_only":"0","restrict_role":"0","restrict_capability":"0","hidden":"0","read_only":"0","roles_allowed":["administrator"],"rest_read":"0","rest_write":"0","rest_pick_response":"array","rest_pick_depth":"2","pick_select_text":"","pick_table_id":"","pick_table_index":"","pick_display":"","pick_where":"","pick_orderby":"","pick_groupby":""}},"show_in_menu":"1","label_singular":"Specialità","public":"1","show_ui":"1","supports_title":"1","supports_editor":"1","publicly_queryable":"1","exclude_from_search":"0","capability_type":"post","capability_type_custom":"specialit","capability_type_extra":"1","has_archive":"0","hierarchical":"0","rewrite":"1","rewrite_with_front":"1","rewrite_feeds":"0","rewrite_pages":"1","query_var":"1","can_export":"1","default_status":"draft","supports_author":"0","supports_thumbnail":"0","supports_excerpt":"0","supports_trackbacks":"0","supports_custom_fields":"0","supports_comments":"0","supports_revisions":"0","supports_page_attributes":"0","supports_post_formats":"0","built_in_taxonomies_category":"0","built_in_taxonomies_elementor_font_type":"0","built_in_taxonomies_elementor_library_type":"0","built_in_taxonomies_link_category":"0","built_in_taxonomies_post_tag":"0","show_in_nav_menus":"1","show_in_admin_bar":"1","pfat_enable":"0","pfat_run_outside_loop":"0","pfat_append_single":"append","pfat_filter_single":"the_content","pfat_append_archive":"append","pfat_filter_archive":"the_content","rest_enable":"0","read_all":"0","write_all":"0"},"6":{"id":6,"name":"studi","label":"Studi","description":"","type":"post_type","storage":"meta","object":"","alias":"","fields":{"indirizzo":{"id":7,"name":"indirizzo","label":"Indirizzo","description":"","help":"","class":"","type":"text","weight":0,"pick_object":"","pick_val":"","sister_id":"","required":"0","text_allow_shortcode":"0","text_allow_html":"0","text_allowed_html_tags":"strong em a ul ol li b i","text_max_length":"255","pick_post_status":["publish"],"admin_only":"0","restrict_role":"0","restrict_capability":"0","hidden":"0","read_only":"0","roles_allowed":["administrator"],"rest_read":"0","rest_write":"0","rest_pick_response":"array","rest_pick_depth":"2","unique":"0","text_repeatable":"0","text_placeholder":""},"telefono":{"id":8,"name":"telefono","label":"Telefono","description":"","help":"","class":"","type":"text","weight":1,"pick_object":"","pick_val":"","sister_id":"","required":"0","text_allow_shortcode":"0","text_allow_html":"0","text_allowed_html_tags":"strong em a ul ol li b i","text_max_length":"255","pick_post_status":["publish"],"admin_only":"0","restrict_role":"0","restrict_capability":"0","hidden":"0","read_only":"0","roles_allowed":["administrator"],"rest_read":"0","rest_write":"0","rest_pick_response":"array","rest_pick_depth":"2","unique":"0","text_repeatable":"0","text_placeholder":""},"avvocati":{"id":9,"name":"avvocati","label":"Avvocati","description":"","help":"","class":"","type":"pick","weight":2,"pick_object":"post_type","pick_val":"avvocati","sister_id":"14","required":"0","pick_format_type":"single","pick_format_single":"dropdown","pick_format_multi":"checkbox","pick_allow_add_new":"1","pick_taggable":"0","pick_show_icon":"1","pick_show_edit_link":"1","pick_show_view_link":"1","pick_limit":"0","pick_user_role":[],"pick_post_status":["publish"],"admin_only":"0","restrict_role":"0","restrict_capability":"0","hidden":"0","read_only":"0","roles_allowed":["administrator"],"rest_read":"0","rest_write":"0","rest_pick_response":"array","rest_pick_depth":"2","unique":"0","pick_select_text":"","pick_table_id":"","pick_table_index":"","pick_display":"","pick_where":"","pick_orderby":"","pick_groupby":""}},"show_in_menu":"1","label_singular":"Studio","public":"1","show_ui":"1","supports_title":"1","supports_editor":"1","publicly_queryable":"1","exclude_from_search":"0","capability_type":"post","capability_type_custom":"studi","capability_type_extra":"1","has_archive":"0","hierarchical":"0","rewrite":"1","rewrite_with_front":"1","rewrite_feeds":"0","rewrite_pages":"1","query_var":"1","can_export":"1","default_status":"draft","supports_author":"0","supports_thumbnail":"0","supports_excerpt":"0","supports_trackbacks":"0","supports_custom_fields":"0","supports_comments":"0","supports_revisions":"0","supports_page_attributes":"0","supports_post_formats":"0","built_in_taxonomies_category":"0","built_in_taxonomies_elementor_font_type":"0","built_in_taxonomies_elementor_library_type":"0","built_in_taxonomies_link_category":"0","built_in_taxonomies_post_tag":"0","show_in_nav_menus":"1","show_in_admin_bar":"1","pfat_enable":"0","pfat_run_outside_loop":"0","pfat_append_single":"append","pfat_filter_single":"the_content","pfat_append_archive":"append","pfat_filter_archive":"the_content","rest_enable":"0","read_all":"0","write_all":"0"}}}

**Comment by jimtrue** [comment]
@gdeldeo Did you open a ticket with Elementor Pro as well? If so can you provide us the link to that bug submission? Thanks!

**Comment by PortalPacific** [comment]
Hi Jim! I am having the same issue. He is actually referring to this plugin: https://wordpress.org/plugins/ele-custom-skin/

The above plugin adds an additional "template" option to create custom loop designs. Elementor is lacking the ability to edit your archive templates, etc. This allows you to do that while pulling in the loop of your choice anywhere you like with Elementor. 

https://www.youtube.com/watch?v=NPkL7AesG-E - here is the basic setup & how it works.

While elementors "dynamic" options work in these loops for creating custom archive templates. A pods "element" is unable to pull in any data from the post(s).

I also opened a support issue with the plugin here: https://wordpress.org/support/plugin/ele-custom-skin

However it isnt listed yet because my posts are being "held for moderation" but i imagine should be up soon.

Thanks for any help!

**Comment by jimtrue** [comment]
Pretty much on this one, have to find out where the 'lines' are drawn as in which plugin is actually handling the output. Is it Elementor Pro and their integration or the Elementor Custom Skin integration.

We're happy to help in anyway we can, but we didn't write either of those integrations and would need their authors to step in to correct any issues in their plugins.

**Comment by PortalPacific** [comment]
I have notified the Custom Skin plugin support of the issue and asked them to chime in, in this discussion. (linked to support topic above)

Hopfully we can come up with a solution as it would open up a lot of doors for many people when customizing their sites.

Thanks for any attention to this!

**Comment by PortalPacific** [comment]
So just a quick re-crap.

I have this Shortcode in a pods template.

```
[dyn-play-add-button mp3_src="{@mp3_url}" title="{@post_title}" artist="The More Perfect Union Podcast" album="Real Debate Without The Hate" date="{@post_date}" image="{@post_thumbnail_url.medium}"][/dyn-play-add-button]
```

 It works great throughout the site on single pages and posts. However when added to a loop or archive template, it doesn't work and shows the error. When checking the box for the "Use current post (singular) or term (term archive)" option. the error  "POD Not Found error". is displayed

As a temporary work around I had to make a custom Elementor widget plugin which inserts the shortcode and pulls in the data manually. Like so.

```
protected function render() {
		global $post;
		$mp3_url = get_post_meta($post->ID, 'mp3_url', true);
		$post_title = $post->post_title;
		$post_date = $post->post_date;
		$post_thumbnail_url = get_the_post_thumbnail_url($post->ID, 'medium');
		$settings = $this->get_settings_for_display();

		echo do_shortcode('[dyn-play-add-button mp3_src="' . $mp3_url . '" title="' . $post_title . '" artist="The More Perfect Union Podcast" album="Real Debate Without The Hate" date="' . $post_date . '" image="' . $post_thumbnail_url . '"][/dyn-play-add-button]');
	}
```

This is part of a custom Elementor widget plugin which i just use instead of the pods "single item" widget in Elementor.

Excited for the response from the Elementor custom skin plugin devs.

Thanks!

**Comment by dudaster** [comment]
The way the plugin works is simple:
- gets an elementor template rendered with elementor function (no custom build)
- puts inside an article the content.

So everything it's done by elementor. It's basically the same as using a template as a shortcode or something like that. The difference is that for each content the post changes.

**Comment by jimtrue** [comment]
@PortalPacific @dudaster Have we heard any feedback from either of the other plugins on this one? If not, we're closing this as an Edge Case. 

@PortalPacific it sounds like you were able to create a Workaround and that will have to do until the premium plugins involved in this issue can provide feedback.

**Comment by Arohonka** [comment]
I found a way to show relations image from post in post query, just add this to elementor text widget:
[pods name="name of main post" slug="ID of relations field"]
[each name of relations field]
{@relations.post_thumbnail.large}
[/each]
[/pods]

another example:
[pods name="show" slug="2377"]
[each model]
{@model.post_thumbnail.large}
[/each]
[/pods]

**Comment by KimchaC** [comment]
I spent all day fighting this same issue and finally found a super easy solution.

I am documenting the solution here for anyone else who finds this through google.

There is no bug or problem. Neither in pods, nor in Elementor or Elementor Custom Skins.

It’s just a usage issue.

**TLDR – The solution**
Most Elementor fields are not shortcode enabled and won’t render them at the right time.

But any field that has the “Dynamic” function can use the Site -> Shortcode option to execute the shortcode at exactly the right time.

Using this option solves pods shortcodes not working on Elementor custom post type templates and in elementor custom skin loops.

So for the detailed explanation for people who care…

**What problem were we trying to solve?**
Display a pods field inside an Elementor widget like the heading widget.

This can be done easily with Elementor’s native custom field integration (Dynamic -> Post Custom Field) or through the pods integration (Dynamic -> Pods Field).

But those integrations don’t support getting a field from another post through a relationship.

For that you have to use the pods shortcode [pods field="product.title_short"].

This is where the issues start with Elementor post templates and the Elementor Custom Skins plugin.

So, what’s the problem

There can be two issues:
**1. Issue – The shortcode must be executed at the “right” time**

This results in no output or “Pod not found” when you use the pods shortcode.

Pods requires that the shortcode is executed within the query context of the custom post.

If you use Elementor Custom Skins and just enter the shortcode in the text field of an element it might get executed at the wrong time. It seems like the skin doesn’t execute the shortcodes inside the text field as it loops through the posts (wp_query results).

Instead Elementor seems to execute the shortcode AFTER the skin has generated the HTML for output.

At that point the pods shortcode is running under the context of the primary post you are editing and not the looped content and therefore not finding the custom field and not displaying anything (or displaying an error).

**2. Issue – Not all elements are shortcode enabled**
This results in the shortcode CODE and not CONTENT being output on the page.

It turns out that most elementor fields are not designed to accept and render shortcuts.

It seems to work most of the time, since Elementor probably executes shortcodes in the entire generated HTML before output.

But it doesn’t do it always. For example, when you create a page template it doesn’t seem to do it and will just display the output of the shortcode.

Both can be solved with the solution above.

**Comment by kingfisher64** [comment]
Thank you @infused-kim your advice was very helpful and it is indeed just a usage issue of how to return what's needed. 

**Usage: Adding a "brief description" PODS field to a Woocommerce Product Archive loop using Elementor Custom Skin.**

In order to find the paragraph field add "Post Info widget to your elementor custom skin loop" > remove date/time/author field options and select "custom" from the dropdown and choose "pods field" your textual field will show as an option.
